### PR TITLE
Use dw dynamic config

### PR DESCRIPTION
**Pull request by leonorfmartins**
**What is this feature?**

This feature reads the config which is being set by HGAPI in https://github.com/grafana/hosted-grafana/pull/5725.
The goal is to use the same config of type `map[string]DualWriterMode` into the 3 cases where we need to enable unified storage or the K8s API:
- enable the FE to use the resource K8s API
- BE to register the resource K8S API
- enable or not resource storage DB

This PR should be followed by a cleanup one where we remove old feature toggles.

**Why do we need this feature?**

We need to make the process of onboarding a new resource into unified storage easier.

See the general idea in https://github.com/grafana/search-and-storage-team/issues/54

### Test
Tested with an ephemeral instance and playlists are being written to unified storage:
```select * from resource where namespace='stack-4720' and resource='playlists';
+--------------------------------------+------------------+----------------------+-----------+------------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+-----------+
| guid                                 | resource_version | group                | resource  | namespace  | name           | value                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | action | label_set |
+--------------------------------------+------------------+----------------------+-----------+------------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+-----------+
| 6b5b1211-8232-4e80-9d76-829e388b030d |           144786 | playlist.grafana.app | playlists | stack-4720 | cdw38h6k3tnnkd | {"kind":"Playlist","apiVersion":"playlist.grafana.app/v0alpha1","metadata":{"name":"cdw38h6k3tnnkd","namespace":"stack-4720","uid":"4c1eb7ba-a2ed-4c9d-9bc2-8135fe0f5bc6","creationTimestamp":"2024-08-27T12:22:48Z","annotations":{"grafana.app/createdBy":"user:advzxsvdjnbb4c","grafana.app/originHash":"Grafana v11.3.0 (de4c5dccea)","grafana.app/originName":"UI","grafana.app/originPath":"/playlists/new","grafana.app/originTimestamp":"2024-08-27T12:22:47Z"}},"spec":{"title":"leonor2","interval":"5m","items":[{"type":"dashboard_by_uid","value":"advzss6liqe4gb"}]}}
 |      1 | NULL      |
+--------------------------------------+------------------+----------------------+-----------+------------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+-----------+
```

when config is
```
[unified_storage.playlists.playlist.grafana.app]
dualWriterMode = 2
```

It should only be merged after https://github.com/grafana/hosted-grafana/pull/5780

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -618,3 +618,11 @@ func (hs *HTTPServer) registerRoutes() {
 	r.Get("/api/snapshots-delete/:deleteKey", reqSnapshotPublicModeOrSignedIn, routing.Wrap(hs.DeleteDashboardSnapshotByDeleteKey))
 	r.Delete("/api/snapshots/:key", reqSignedIn, routing.Wrap(hs.DeleteDashboardSnapshot))
 }
+
+func (hs *HTTPServer) shouldUseK8sAPI(resource string) bool {
+	unifiedStorageOptions := hs.Cfg.UnifiedStorage
+	if _, ok := unifiedStorageOptions[resource]; ok {
```
Comment (filewalkwithme) [inline pkg/api/api.go:624]: If we want to disable the usage of K8sAPI for a given resource, the only way would be to remove the entire section...

would it make sense to parse a field called `enabled` (default true)?

File: pkg/services/featuremgmt/registry.go
Comment (ryantxu) [inline pkg/services/featuremgmt/registry.go:1]: unrelated?

File: pkg/setting/setting.go [hunk 2]
```diff
@@ -523,6 +524,13 @@ type Cfg struct {
 
 	//Short Links
 	ShortLinkExpiration int
+
+	// Unified Storage
+	UnifiedStorage map[string]UnifiedStorageConfig
```
Comment (ryantxu) [inline pkg/setting/setting.go:529]: frontend really should not know anything about unified storage

**Review by ryantxu — CHANGES_REQUESTED** [review]
on the frontend, i think feature flags are the right thing to do.    Exposing the matrix of options to the frontend will be a mess.

Is the concern here setting mode 3+, but the UI still writing to legacy apis?

**Comment by leonorfmartins** [comment]
hm, so the concern (more like the goal) is to have just one config which can dictate not only what API to expose and use, but also the storage. Would something like
```
UnifiedStorage: extractKeyResourcesFromConfig(hs.Cfg.UnifiedStorage)
```
work?

**Comment by filewalkwithme** [comment]
I am going to address the remaining comments

**Comment by filewalkwithme** [comment]
Just adding to the PR description, the main goal of this PR is to introduce Config INI sections for each Unified Storage resource. That will give us flexibility to better centralize and control the onboarding and rollout of new resources (playlists, banners, etc...). 

This is being done in the same spirit as what happened to Unified Alerting. Initially we had a feature toggle for it (`unified_storage`) which got replace by a proper config section when the feature got GA

**Comment by filewalkwithme** [comment]
> on the frontend, i think feature flags are the right thing to do. Exposing the matrix of options to the frontend will be a mess.

@ryantxu agree with that. We added a commit to pass a single flag to the frontend indicating whether it should use the k8s API or Legacy API (instead of the feature toggle itself, which we want to remove in a follow-up PR)

c/c @leonorfmartins

File: pkg/api/api.go (hunk 1 above)
Comment (filewalkwithme) [reply pkg/api/api.go:624]: Let's add this on a follow-up PR
c/c @leonorfmartins

File: pkg/api/playlist.go [hunk 3]
```diff
@@ -27,7 +27,9 @@ import (
 func (hs *HTTPServer) registerPlaylistAPI(apiRoute routing.RouteRegister) {
 	// Register the actual handlers
 	apiRoute.Group("/playlists", func(playlistRoute routing.RouteRegister) {
-		if hs.Features.IsEnabledGlobally(featuremgmt.FlagKubernetesPlaylists) {
+		// TODO: remove kubernetesPlaylists feature flag
+		if hs.shouldUseK8sAPI(playlistalpha1.PlaylistResourceInfo.GroupResource().String()) ||
```
Comment (filewalkwithme) [inline pkg/api/playlist.go:31]: I wonder if `shouldUseK8sAPI` should/could be refactored to take into the account the cluster wide feature toggle too.

@leonorfmartins what do you think?

Comment (leonorfmartins) [reply pkg/api/playlist.go:31]: Id say, we should have the same rule to enable the API in the BE and in the FE... so if we allow for cluster-wide feature toggles to control enabling the API here, we should also do the same in the FE?
I'm actually less inclined to do this for now, we can always add it later on if we need to?

**Comment by leonorfmartins** [comment]
tested again with the following configs:
```
[unified_storage.playlists.playlist.grafana.app]
dualWriterMode = 2
[unified_storage_mode]
playlist.grafana.app/playlists = 0 --> so we make sure the dual writer is in mode 2 because of the section above
[feature_toggles]
kubernetesPlaylists = false --> so we make sure it's the section above enabling the K8s API
```

added a new playlist called `leonor-test-again` and it worked:
```
k get playlists -n stack-4720
NAME             TITLE               INTERVAL   CREATED AT
adw00w7whyhvkb   leonor-test         5m         2024-08-26T14:47:06Z
cdw38h6k3tnnkd   leonor2             5m         2024-08-27T12:22:48Z
edw6q3py4ngn4b   leonor-test-again   5m         2024-08-28T11:51:04Z
```

File: pkg/setting/setting_unified_storage.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package setting
+
+import (
+	"strings"
+
+	"github.com/grafana/grafana/pkg/apiserver/rest"
+)
+
+func (cfg *Cfg) setUnifiedStorageConfig() {
```
Comment (ryantxu) [inline pkg/setting/setting_unified_storage.go:9]: Can you add a comment here for what the settings look like in config.ini?

Currently I have:
```
[grafana-apiserver]
storage_type = unified ; 

; Dual write settings
dashboards.dashboard.grafana.app=1
playlists.playlist.grafana.app=2
```

Comment (leonorfmartins) [reply pkg/setting/setting_unified_storage.go:15]: so, config.ini looks like this:
```
[unified_storage.playlists.playlist.grafana.app]
dualWriterMode = 2
```
we can add more fields in each section. That way we can change each config per resource.

**Review by ryantxu — APPROVED** [review]
LGTM
